Book.java

```diff
@@ -0,0 +1,315 @@
+                break;
+
+            case EQUITY:
+                equityAccounts.add(account);
+
+            // Not sure what to do with INCOME, EXPENSES
+        }
+
+    }
+
+    /**
+     * Adding an asset means debiting the account relevant to that type of contract
+     * and crediting equity.
+     * @param contract an Asset contract to add
+     */
+    public void addAsset(Contract contract) {
+        Account assetAccount = assetAccountsMap.get(contract.getClass());
+
+        if (assetAccount==null) {
+            // If there doesn't exist an Account to hold this type of contract, we create it
+            assetAccount = new Account(contract.getClass().getName(), AccountType.ASSET);
+            addAccount(assetAccount, contract.getClass());
+        }
+
+        // (dr asset, cr equity)
+        assetAccount.debit(contract.getValue());
+        equityAccount.credit(contract.getValue());
+
+        // Add the contract to the account's inventory
+        assetAccount.addContract(contract);
+    }
+
+    /**
+     * Adding a liability means debiting equity and crediting the account
+     * relevant to that type of contract.
+     * @param contract a Liability contract to add
+     */
+    public void addLiability(Contract contract) {
+        Account liabilityAccount = liabilityAccountsMap.get(contract.getClass());
+
+
+        if (liabilityAccount==null) {
+            // If there doesn't exist an Account to hold this type of contract, we create it
+            liabilityAccount = new Account(contract.getClass().getName(), AccountType.LIABILITY);
+            addAccount(liabilityAccount, contract.getClass());
+        }
+
+        // (dr equity, cr liability)
+        equityAccount.debit(contract.getValue());
+        liabilityAccount.credit(contract.getValue());
+
+        liabilityAccount.addContract(contract);
+    }
+
+    public void addCash(double amount) {
+
+        // (dr cash, cr equity)
+        cashAccount.debit(amount);
+        equityAccount.credit(amount);
+    }
+
+
+    /**
+     * Operation to cancel a Loan to someone (i.e. cash in a Loan in the Assets side).
+     *
+     * I'm using this for simplicity but note that this is equivalent to selling an asset.
+     * @param amount the amount of loan that is cancelled
+     */
+    public void pullFunding(double amount) {
+        Account loanAccount = assetAccountsMap.get(Loan.class);
+
+        // (dr cash, cr asset )
+        cashAccount.debit(amount);
+        loanAccount.credit(amount);
+
+    }
+
+    /**
+     * Operation to pay back a liability loan; debit liability and credit cash
+     * @param amount amount to pay back
+     */
+    public void payLiability(double amount, Class<? extends Contract> liabilityType) {
+        Account liabilityAccount = liabilityAccountsMap.get(liabilityType);
+
+        //Todo: What do we do if we can't pay??!! At the moment I'm calling my owner to raise liquidity
+        if (cashAccount.getBalance() < amount) {
```

Comment:
@prauwolf Agreed.  I don't know much about the Java conventions regarding checked vs unchecked exceptions, etc.  I think we need some kind of `LiquidityException` that will be thrown in the event that an agent does not have liquid resources available to make required payments.  Ideally this would also be expressed in the return type of the method (which would force the caller of the method to explicitly deal with the possibility that an exception would be returned.